src/Bindings/LuaStateParams.h

```diff
@@ -0,0 +1,571 @@
+	}
+
+
+	/** Terminator for the template-based recursion of the function above - for a single overload. */
+	template <typename... T1>
+	static int ReadInternal(cLuaState & a_LuaState, int a_CurOverload, std::tuple<T1...> & a_Ovl)
+	{
+		// Try to read this overload
+		if (ReadSingleOverload(a_LuaState, a_Ovl))
+		{
+			return a_CurOverload;
+		}
+
+		// No more overloads, report failure
+		return -1;
+	}
+
+
+
+	/** Helper struct to implement iterating over std::tuple elements */
+	template <int N> struct SizeT {};
+
+
+
+	/** Attempts to match the params on the Lua stack to the given API function overload.
+	Returns true if successful, false on failure. */
+	template <typename... T>
+	static bool ReadSingleOverload(cLuaState & a_LuaState, std::tuple<T...> & a_Overload)
+	{
+		// Check that there exactly as many params as the tuple items:
+		if (!lua_isnone(a_LuaState, sizeof...(T) + 1))
+		{
+			// Too many params
+			return false;
+		}
+		if (lua_isnone(a_LuaState, sizeof...(T)))
+		{
+			// Too few params
+			return false;
+		}
+
+		// Read the tuple, compile-time-recursively:
+		return ReadSingleOverloadRecurse(a_LuaState, std::forward<std::tuple<T...>>(a_Overload), SizeT<sizeof...(T)>());
+	}
+
+
+
+	/** Attempts to match the params on the Lua stack to the given API function overload.
+	The compile-time recursive worker implementation of ReadSingleOverload, recurses by the number of elements in the overload tuple.
+	Returns true on success, false on failure. */
+	template <typename... T, size_t N>
+	static bool ReadSingleOverloadRecurse(cLuaState & a_LuaState, std::tuple<T...> & a_Overload, SizeT<N>)
+	{
+		// First read the params from the lower tuple indices:
+		if (!ReadSingleOverloadRecurse(a_LuaState, a_Overload, SizeT<N - 1>()))
+		{
+			return false;
+		}
+
+		// Then read the param pointed to by our index:
+		return GetStackValue(a_LuaState, N, std::get<N - 1>(a_Overload));
+	}
+
+
+	/** Terminator for the above compile-time-recursive function. */
+	template <typename... T>
+	static bool ReadSingleOverloadRecurse(cLuaState & a_LuaState, std::tuple<T...> & a_Overload, SizeT<1>)
+	{
+		return GetStackValue(a_LuaState, 1, std::get<0>(a_Overload));
+	}
+
+
+
+	/** Reads one value from the Lua stack.
+	Returns true on success, false on failure.
+	The hard work is delegated into cLuaState that already has this function, but we need to specialize it for decorators (cNonNil, cSelf etc.). */
+	template <typename T>
+	static bool GetStackValue(cLuaState & a_LuaState, int a_StackPos, T && a_ReturnedVal)
+	{
+		return a_LuaState.GetStackValue(a_StackPos, std::forward<T>(a_ReturnedVal));
+	}
+
+
+	/** Reads one value from the Lua stack.
+	Returns true on success, false on failure.
+	Specialization: Retrieves a value that should represent a non-nil pointer. Used primarily for ReadParams(... nonNil(...)) */
+	template <typename T>
+	static bool GetStackValue(cLuaState & a_LuaState, int a_StackPos, cNonNil<T> & a_ReturnedVal)
```

Comment:
Shouldn't a `cNonNil` be taken by value as it is itself a reference?

Reply:
It's just to cover all cases. Not sure if MSVC actually doesn't try to match a reference and won't accept a value.